lib/ABI/FunctionType/Conversion.cpp

```diff
@@ -375,38 +375,63 @@ TCC::tryConvertingReturnValue(const ReturnValueRegisters &Registers) {
                                   {});
     }
   } else {
-    // Multiple registers mean that it should probably be a struct.
-    //
-    // \note: it's also possible that it's just a big primitive,
-    // \todo: look into supporting those.
-    auto [ReturnType, ReturnTypePath] = Bucket.makeType<model::StructType>();
-    for (auto Register : Ordered) {
-      // Make a separate field for each register.
-      model::StructField Field;
-      Field.Offset() = ReturnType.Size();
-      if (auto Iter = Registers.find(*Ordered.begin()); Iter != Registers.end())
-        Field.Type() = Iter->Type();
-      else
-        Field.Type() = { Bucket.genericRegisterType(*Ordered.begin()), {} };
-
-      auto FieldSize = Field.Type().size();
-      revng_assert(FieldSize.has_value() && FieldSize.value() != 0);
+    // Multiple registers, it's either a struct or a big scalar.
 
-      // Round the next offset based on the natural alignment.
-      auto Alignment = ABI.alignment(Field.Type());
-      revng_assert(Alignment.has_value() && Alignment.value() != 0);
-      ReturnType.Size() += (*Alignment - ReturnType.Size() % *Alignment);
+    // First, check if it can possibly be a struct.
+    if (ABI.MaximumGPRsPerAggregateReturnValue() < Ordered.size()) {
+      // It cannot be a struct, it's bigger than allowed.
+      if (ABI.MaximumGPRsPerScalarReturnValue() < Ordered.size()) {
+        // It cannot be a scalar either.
+        revng_log(Log,
+                  "No known return value type supports that many registers ("
+                    << Ordered.size() << ") under "
+                    << model::ABI::getName(ABI.ABI()) << ":\n");
+        return std::nullopt;
+      }
 
-      // Insert the field
-      ReturnType.Fields().insert(std::move(Field));
+      // It's probably a scalar, replace its type with a fake one making sure
+      // at least the size adds up.
+      //
+      // TODO: sadly this discards type information from the registers, look
+      //       into preserving it at least partially.
+      std::size_t PointerSize = model::ABI::getPointerSize(ABI.ABI());
+      return model::QualifiedType{
+        Bucket.getPrimitiveType(model::PrimitiveTypeKind::Values::Generic,
+                                PointerSize * Ordered.size()),
+        {}
+      };
+    } else {
+      // It could be either a struct or a scalar, go the concervative route
```

Comment:
Never see any indication anything supports scalars with size exceeding two registers. To be extra pedantic, `abi::Definition::MaximumGPRsPerScalarReturnValue` (or `...Argument`, but it's irrelevant here) is what dictates the maximum size it can be. I didn't see much reason to lock it to two: it doesn't change the complexity much.

Reply:
And also a bit of insight why we need to handle this explicitly.

Consider x86 SystemV abi, it doesn't allow returning structs in registers, only scalars, so if we see any registers straight-up returned (non-`SPTAR`-style), we know for sure it's a scalar - hence we handle it as such.

Reply:
The comment says:

> It could be either a struct or a scalar

In which situation we end up here with a scalar? If never, amend the comment.

Reply:
Wait, I think you're misunderstanding the controlflow here. It's going something like
```cpp
if (not canItBeAStruct(/* whatever*/) {
  if (not canItBeAScalar(/* whatever */) {
    return std::nullopt; // the ABI's probably not right.
  }
  
  return emitScalar(/* whatever */);
} else {
  // It could be either a struct or a scalar (we have no way of knowing which), so we emit a struct.
  return emitStruct(/* whatever */);
}
```

Reply:
TL;DR: it doesn't have to be BIG to be able to reach that part, it just need to occupy more than a single register.

Reply:
Ah, I probably misread.